src/samples/network/echo-client.c

```diff
@@ -0,0 +1,178 @@
+ * To see the usage help, -h or --help.
+ */
+
+#include <errno.h>
+#include <getopt.h>
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+
+#include "soletta.h"
+#include "sol-buffer.h"
+#include "sol-socket.h"
+#include "sol-network.h"
+#include "sol-vector.h"
+#include "sol-util.h"
+
+static struct sol_socket *sock;
+static struct sol_network_link_addr address;
+
+static bool
+on_can_read(void *data, struct sol_socket *s)
+{
+    ssize_t r;
+    char buffer[1024] = { };
+    struct sol_network_link_addr addr;
+
+    r = sol_socket_recvmsg(s, buffer, 1024, &addr);
+    if (r < 0) {
+        fprintf(stderr, "ERROR: Failed in receiving the message\n");
+        goto err;
+    }
+
+    printf("Received: %s\n", buffer);
+    sol_quit_with_code(EXIT_SUCCESS);
+
+    return true;
+
+err:
+    sol_quit_with_code(EXIT_FAILURE);
+    return false;
+}
+
+static bool
+on_can_write(void *data, struct sol_socket *s)
+{
+    int r = -1;
+    const char *value = data;
+
+    r = sol_socket_sendmsg(sock, value, strlen(value), &address);
+    if (r < 0) {
+        fprintf(stderr, "ERROR: Could not send data\n");
+        goto end;
+    }
+
+    r = 0;
+
+end:
+    return (r < 0) ? false : true;
+}
+
+static void
+startup_client(void)
+{
+    char **argv = sol_argv();
+    char *itr, *addr = NULL;
+    char addr_str[PATH_MAX] = { };
+    size_t len;
+    int c, opt_idx, argc = sol_argc();
+    static const struct option opts[] = {
+        { "address", required_argument, NULL, 'a' },
+        { "help", no_argument, NULL, 'h' },
+        { 0, 0, 0, 0 }
+    };
+    struct sol_socket_ip_options options = {
+        .base  = {
+            SOL_SET_API_VERSION(.api_version = SOL_SOCKET_OPTIONS_API_VERSION, )
+            SOL_SET_API_VERSION(.sub_api = SOL_SOCKET_OPTIONS_API_VERSION, )
+            .on_can_write = on_can_write,
+            .on_can_read = on_can_read,
+            .family = SOL_NETWORK_FAMILY_INET6,
+        },
+        .secure = false,
+        .reuse_addr = true,
+    };
+
+    while ((c = getopt_long(argc, argv, "a:h", opts, &opt_idx)) != -1) {
+        switch (c) {
+        case 'a':
+            addr = optarg;
+            break;
+        case 'h':
+        default:
+            fprintf(stderr,
+                "Usage:\n\t%s -a <ip-port> value\n", argv[0]);
+            sol_quit_with_code(EXIT_SUCCESS);
+            return;
+        }
+    }
+
+    if (!addr) {
+        fprintf(stderr, "ERROR: No address was given\n");
+        goto err;
+    }
+
+    options.base.data = argv[optind];
+    if (!options.base.data) {
+        fprintf(stderr, "ERROR: No data was given, use -h for help\n");
+        goto err;
+    }
+
+    sock = sol_socket_new((struct sol_socket_options *)&options);
+    if (!sock) {
+        fprintf(stderr, "ERROR: Could not create the socket\n");
+        goto err;
+    }
+
+    address.family = SOL_NETWORK_FAMILY_INET6;
+
+    itr = strstr(addr, "-");
```

Comment:
also `-` is not standard to split port and address. Usually it is `':'`.

Then `echo` service has a specific port, 7 as per `/etc/services`. It would be nice to default to that and also implement the actual echo service protocol (if there is anything special)